File: ts/src/service-registry.ts
import { type Environment, type EnvironmentPreset, ENVIRONMENTS } from './environments.js'

export interface ServiceUrls {
  /** Gateway fallback — dùng cho health, license, và các endpoint chưa migrate */
  gateway: string
  /** channel-service — version (v1/v2/v3) thêm trong từng method */
  channelService: string
  /** platform — version (v1/v2) thêm trong từng method */
  platform: string
  /** ips/v1 — develop: ips.dev.imbrace.lan/ips/v1, các env khác: gateway/ips/v1 */
  ips: string
  /** data-board — không có version prefix, path trực tiếp */
  dataBoard: string
  /** ai — version (v2/v3) thêm trong từng method */
  ai: string
  /** marketplaces/v1 — standalone marketplace service */
  marketplaces: string
}

export function resolveServiceUrls(
  env: Environment | EnvironmentPreset,
  overrides?: Partial<ServiceUrls>,
): ServiceUrls {
  const preset: EnvironmentPreset =
    typeof env === 'string' ? ENVIRONMENTS[env] : env
  const gw    = preset.gateway.replace(/\/$/, '')
  const hosts = preset.serviceHosts ?? {}

  const resolved: ServiceUrls = {
    gateway:        gw,
    channelService: `${gw}/channel-service`,
    platform:       `${gw}/platform`,
    ips:            `${(hosts.ips ?? gw).replace(/\/$/, '')}/ips/v1`,
    dataBoard:      `${(hosts.dataBoard ?? gw).replace(/\/$/, '')}/data-board`,
    ai:             `${gw}/ai`,
    marketplaces:   `${gw}/marketplaces`,
  }

  return { ...resolved, ...overrides }
}

File: py/src/imbrace/service_registry.py
from __future__ import annotations
from typing import Optional, Union
from dataclasses import dataclass

from .environments import EnvironmentPreset, ENVIRONMENTS


@dataclass
class ServiceUrls:
    gateway: str
    channel_service: str
    platform: str
    ips: str
    data_board: str
    ai: str
    marketplaces: str


def resolve_service_urls(
    env: Union[str, EnvironmentPreset],
    overrides: Optional[dict] = None,
) -> ServiceUrls:
    if isinstance(env, str):
        preset = ENVIRONMENTS[env]
    else:
        preset = env

    gw = preset.gateway.rstrip("/")

    resolved = ServiceUrls(
        gateway=gw,
        channel_service=f"{gw}/channel-service",
        platform=f"{gw}/platform",
        ips=f"{(preset.service_hosts.ips or gw).rstrip('/')}/ips/v1",
        data_board=f"{(preset.service_hosts.data_board or gw).rstrip('/')}/data-board",
        ai=f"{gw}/ai",
        marketplaces=f"{gw}/marketplaces",
    )

    if overrides:
        for key, value in overrides.items():
            if hasattr(resolved, key) and value is not None:
                setattr(resolved, key, value)

    return resolved
